### DISCUSSION TITLE
first time running airflow CLI does not create airflow.cfg in custom $AIRFLOW_HOME

### DISCUSSION THREAD

**Discussion by sswander** [apache/airflow#33951] [General] [answered]
### Apache Airflow version

2.7.0

### What happened

Using Airflow 2.7.0, running the following command does not result in airflow.cfg being created
```sh
$ AIRFLOW_HOME=~/airflow_test airflow --help

$ ls ~/airflow_test
logs # <-- should include airflow.cfg too
```

**Further background info**
Our 3rd party tool invokes `airflow --help` to create a airflow.cfg file into a custom AIRFLOW_HOME env.

### What you think should happen instead

The 2.7.0 docs mentioned that it will create airflow.cfg in custom AIRFLOW_HOME https://airflow.apache.org/docs/apache-airflow/2.7.0/howto/set-config.html

Therefore, 

### How to reproduce

Using virtualenv to conveniently reproduce:

```sh
virtualenv test_airflow_270
. test_airflow_270/bin/activate

(test_airflow_270) -> ~ pip install apache-airflow=2.7.0
(test_airflow_270) -> ~ AIRFLOW_HOME=~/airflow270_test_home airflow --help
(test_airflow_270) -> ~ ls ~/airflow270_test_home # -> does not contain airflow.cfg
```

### Operating System

macOS Ventura `13.5.1 (22G90)`

### Deployment

Virtualenv installation

### Anything else

This behaved according to the docs in Airflow 2.6.3, but in Airflow 2.7.0 it stopped working

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by sswander** [comment]
Hi there! I'm willing to raise PR but I'm very new to Airflow and python, and after spending some time reading the source code, I could not trace where this `airflow.cfg` is created. If there is any pointer, I would gladly follow it up 🙇

**Reply by Taragolis** [reply]
`airflow.cfg` created in the first access to the configuration and `--help` do not spawn it because it only show available command in airflow cli. 

If you run something like `AIRFLOW_HOME=~/airflow270_test_home airflow config list` it would be generated.

**Reply by Taragolis** [reply]
And additional for the second question, where it templates stored and which part of airflow create it

1. Templates stored into the [airflow/config_templates](https://github.com/apache/airflow/tree/faa8f543d5f7bf7113076dab5f579493ec2860ec/airflow/config_templates)  

2. Default file created by this function
https://github.com/apache/airflow/blob/b82ce61285f3f4f0c7eccb2c3effaef53c9fb84e/airflow/configuration.py#L1976-L1999

3. Which only runs for certain [airflow cli](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html) commands
https://github.com/apache/airflow/blob/5e1e5fa637aa90b92e4b0e1d62c577461ea92369/airflow/__main__.py#L49-L55

**Reply by sswander** [reply]
Thanks for the tip @Taragolis ! I can confirm that `AIRFLOW_HOME=~/airflow270_test_home airflow config list` works 👍 

However I'm also checking because `airflow --help` used to create `airflow.cfg` in 2.6.3, and that is no longer the case in 2.7.0. Is this something that is intended? If it's not intended, I can help put back this behaviour if I get some pointers on where this creation happens in the code base. If this behaviour change is intended, then all is well, I'll close this discussion.

EDIT: whoops seems I replied too quickly. I saw your additional reply. Will study these. Thanks a lot!

**Reply by Taragolis** [reply]
I have a look, this behaviour was added here: https://github.com/apache/airflow/pull/32765
Even if it not mentioned in the PR but IMHO `--help` should only show informations about available commands/sub-comands.

**Reply by Taragolis** [reply]
@sswander And I've just wondering this previous behaviour of `airflow --help` it is personal finding, somewhere mention into the documentation, or just recommendation which are posted in other resource?

**Reply by sswander** [reply]
Hey @Taragolis, thanks for the link to the Github PR, that's really helpful 🙇 

This behaviour of `airflow --help` is discovered from here https://github.com/meltano/meltano/blob/e4bdaedab02462e9e19a1bf063cbce26bc3c7581/src/meltano/core/plugin/airflow.py#L114-L120

We're a user of Meltano, which integrates airflow to our ELT pipeline. Ever since upgrading to airflow 2.7.0, our ELT pipeline crashes, and while trying to debug, we identified this behaviour change as a root cause.

The first instinct was to restore the behaviour to 2.7.0 (if intended). Otherwise, like your recommendation, Meltano probably has to use another command to generate `airflow.cfg` that is not `airflow --help`.

**Reply by Taragolis** [reply]
> The first instinct was to restore the behaviour to 2.7.0 (if intended)

Let's  try gently ping Jarek @potiuk, I guess he know should we return this behaviour or not, and reason

> We're a user of Meltano, which integrates airflow to our ELT pipeline. Ever since upgrading to airflow 2.7.0, our ELT pipeline crashes, and while trying to debug, we identified this behaviour change as a root cause.

I would recommend also open an issue ticket in Meltano

BTW `airflow.cfg` it is not only single source of configs:
Another place is environment variables AIRFLOW__{SECTION_NAME}__{CONFIG_PARAMETER}, which have a higher priority over airflow.cfg

**Reply by potiuk** [reply, marked as answer]
Yes. We deliberatly do NOT create the config when you run `--help`. The fact that `--help` was doing it, was purely accidental. 

Producing config when you use help really makes no sense whatsoever - it makes sense to produce config when you want to run command that will use that config, but not when you run `--help` to get the list of commands to run.

Instead airflow 2.7.0 got a much more comprehensive `airflow config` command that allows you to not only generate the config as you want and with as many details as you want, but also it behaves predictably. Previously wrting the configuration automatically was behaving in a way that was not really described well - t was not described what kind of values you will get when you run config (defaults? defaults with current configuration taken from variables?). 

Sometimes it was generated, and sometimes it was not.

The new command will return:

* default values when you use `--defaults`
* values calculated from "current" configuration + defaults when you do 

And you are in full control of what and where to write. There are more configuration options you can choose from (see the CLI help). 

For example:

https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html

```
However, for production case you are advised to generate the configuration using command line:

airflow config list --defaults
This command will produce the output that you can copy to your configuration file and edit.

It will contain all the default configuration options, with examples, nicely commented out so you need only un-comment and modify those that you want to change. This way you can easily keep track of all the configuration options that you changed from default and you can also easily upgrade your installation to new versions of Airflow when they come out and automatically use the defaults for existing options if they changed there.

You can redirect it to your configuration file and edit it:

airflow config list --defaults > "${AIRFLOW_HOME}/airflow.cfg"

```

You should use this command instead. You might also want to run `airflow version` to check if airflow version is < 2.7.0 and run the pevious "accidental" behaviour from pre 2.7.0.

And yes this is typical case of Hyrum's law in practice:

![image](https://github.com/apache/airflow/assets/595491/83df2d62-e799-4ed8-9291-3eac1a1f5a4a)

**Reply by potiuk** [reply]
Of course that's a bit of exxageration, but yes. you apparently relied on undocumented and accidental behaviour.

**Reply by potiuk** [reply]
And yes. I recommend to open an issue to meltano and point them to this thread.

**Reply by potiuk** [reply]
Also (in case you or someone from Meltano wants to know WHY it is like that). 

Because of SPEED. 

In order to generate default configuration for airflow after https://github.com/apache/airflow/pull/32604 - we need to not only look at "airflow" configuration but also at all "provider" configuration. This means we need to discover all installed providers (via entrypoints) and read from their get_provider_info entrypoints, information on what kind of configuration each provider contributes. This - when you have big number of providers - takes couple of hundred milliseconds - sometimes more a second. Which would mean that if you would like to generate configuration when running `--help` or `version` the first time you run airflow cli (those two commands don't generate configuration because they are purely informational)  you would have to wait about a second or so to see list of commands or version of airflow.

Therefore we have now comprehensive command that is dedicated to that, and neither --help nor version will not be slowed down by the need to discover providers.

**Reply by sswander** [reply]
Hey, thanks so much for sharing these information @potiuk ! It's clear to me now that the previous behaviour was accidental. In this case it makes a lot of sense to open an issue on Meltano, which I will do. Also I appreciate the time taken to explain the details; I learnt more about Airflow's configuration from this conversation 👍
